README
CRONOLOG version 1.6.1

"cronolog" is a simple program that reads log messages from its input
and writes them to a set of output files, the names of which are
constructed using a template and the current date and time.  The
template uses the same format specifiers as the Unix date command
(which are the same as the standard C strftime library function).

"cronolog" is intended to be used in conjunction with a Web server, such
as Apache, to split the access log into daily or monthly logs. For
example the Apache configuration directives:

	TransferLog "|/www/sbin/cronolog /www/logs/%Y/%m/%d/access.log"
	ErrorLog    "|/www/sbin/cronolog /www/logs/%Y/%m/%d/errors.log"

would instruct Apache to pipe its access and error log messages into
separate copies of cronolog, which would create new log files each day
in a directory hierarchy structured by date, i.e. on 31 December 1996
messages would be written to

	/www/logs/1996/12/31/access.log
	/www/logs/1996/12/31/errors.log

after midnight the files

	/www/logs/1997/01/01/access.log
	/www/logs/1997/01/01/errors.log

would be used, with the directories 1997, 1997/01 and 1997/01/01 being
created if they did not already exist.

Each character in the template represents a character in the expanded
filename, except for date and time format specifiers, which are
replaced by their expansion.  Format specifiers consist of a `%'
followed by one of the following characters:

 %	a literal % character
 n	a new-line character
 t	a horizontal tab character

Time fields:

 H 	hour (00..23)
 I 	hour (01..12)
 p 	the locale's AM or PM indicator
 M 	minute (00..59)
 S 	second (00..61, which allows for leap seconds)
 X 	the locale's time representation (e.g.: "15:12:47")
 Z 	time zone (e.g. GMT), or nothing if the time zone cannot be determined

Date fields:

 a 	the locale's abbreviated weekday name (e.g.: Sun..Sat)
 A 	the locale's full weekday name (e.g.: Sunday .. Saturday)
 b 	the locale's abbreviated month name (e.g.: Jan .. Dec)
 B 	the locale's full month name, (e.g.: January .. December)
 c 	the locale's date and time (e.g.: "Sun Dec 15 14:12:47 GMT 1996")
 d 	day of month (01 .. 31)
 j 	day of year (001 .. 366)
 m 	month (01 .. 12)
 U 	week of the year with Sunday as first day of week (00..53, where
	week 1 is the week containing the first Sunday of the year)
 W 	week of the year with Monday as first day of week (00..53, where
	week 1 is the week containing the first Monday of the year)
 w 	day of week (0 .. 6, where 0 corresponds to Sunday)
 x 	locale's date representation (e.g. today in Britain: "15/12/96")
 y 	year without the century (00 .. 99)
 Y 	year with the century (1970 .. 2038)

Other specifiers may be available depending on the C library's
implementation of the strftime function.  Before writing a message
cronolog checks the time to see whether the current log file is still
valid and if not it closes the current file, expands the template
using the current date and time to generate a new file name, opens the
new file (creating missing directories on the path of the new log file
as needed unless the program is compiled with -DDONT_CREATE_SUBDIRS)
and calculates the time at which the new file will become invalid.

The most up-to-date version of "cronolog" can be found at:

	http://www.ford-mason.co.uk/resources/cronolog/


INSTALLATION
============

Cronolog uses GNU autoconf, so it is configured and built
with the standard sequence:

	./configure
	make

To install type "make install" or simply copy the executable
"src/cronolog" to a suitable directory.

Cronolog has a number of options that can be set either by including
in CFLAGS when configuring, or by editing the file "src/config.h":

	-DFILE_MODE=octal-number	mode used for creating files (default is 0664)
	-DDIR_MODE=octal-number		mode used for creating directories (default is 0775)
	-DDONT_CREATE_SUBDIRS		don't include code to create missing directories
	-DNEED_GETOPT_DEFS		if your platform doesn't declare getopt()

e.g. to set the file mode to 0600:

	CFLAGS="-O2 -DFILE_MODE=0600" ./configure


TESTING
=======

The package includes a simple test harness program (cronotest), which
tests out the lower level functions and prints out debugging
information.  The test program is run with the command line:

	cronotest [-d] "template" count

template is the filename template and count is the number of periods
to evaluate, for example:

	$ cronotest "%Y/%m/%d/access.log" 4
	Determining periodicity of "%Y/%m/%d/access.log"
	%Y -> yearly
	%m -> monthly
	%d -> daily
	Rotation period is per day
	Start time is Sun Dec 15 15:55:43 1996 (850665343)
	Period   1 starts at Sun Dec 15 00:00:00 1996 (850608000):  "1996/12/15/access.log"
	Period   2 starts at Mon Dec 16 00:00:00 1996 (850694400):  "1996/12/16/access.log"
	Period   3 starts at Tue Dec 17 00:00:00 1996 (850780800):  "1996/12/17/access.log"
	Period   4 starts at Wed Dec 18 00:00:00 1996 (850867200):  "1996/12/18/access.log"

Specifying the flag -d instructs the test program to create missing
subdirectories, for example, having created a directory 1996:

	$ cronotest -d  "%Y/%m/%d/access.log" 4
	Determining periodicity of "%Y/%m/%d/access.log"
	%Y -> yearly
	%m -> monthly
	%d -> daily
	Rotation period is per day
	Start time is Sun Dec 15 16:00:21 1996 (850665621)
	Period   1 starts at Sun Dec 15 00:00:00 1996 (850608000):  "1996/12/15/access.log"
	Creating missing components of 1996/12/15/access.log
	Testing directory "1996"
	Testing directory "1996/12"
	Directory "1996/12" does not exist -- creating
	Testing directory "1996/12/15"
	Directory "1996/12/15" does not exist -- creating
	Period   2 starts at Mon Dec 16 00:00:00 1996 (850694400):  "1996/12/16/access.log"
	Creating missing components of 1996/12/16/access.log
	Initial prefix "1996" known to exist
	Initial prefix "1996/12" known to exist
	Testing directory 1996/12/16
	Directory "1996/12/16" does not exist -- creating
	Period   3 starts at Tue Dec 17 00:00:00 1996 (850780800):  "1996/12/17/access.log"
	Creating missing components of 1996/12/17/access.log
	Initial prefix "1996" known to exist
	Initial prefix "1996/12" known to exist
	Testing directory 1996/12/17
	Directory "1996/12/17" does not exist -- creating
	Period   4 starts at Wed Dec 18 00:00:00 1996 (850867200):  "1996/12/18/access.log"
	Creating missing components of 1996/12/18/access.log
	Initial prefix "1996" known to exist
	Initial prefix "1996/12" known to exist
	Testing directory 1996/12/18
	Directory "1996/12/18" does not exist -- creating


Extra debugging code was added to cronolog at version 1.5b8 to
facilitate interactive testing.  To activate this cronolog must be
must be invoked with -x (use -x- to send debug messages to stderr).
In this mode each time it reads a message cronolog will print a
message giving the current time, the start of the next period and the
number of seconds until the start of the next period.  This option is
obviously not very useful when cronolog is being used for a production
server handling several thousands or millions of requests a day, but
can be useful for checking that cronolog is working correctly.  You
could set up a virtual server that was only used for testing and
periodically send it a request.  The -s option allows a start time to
be specified -- cronolog will pretend that that is the current time.

The following transcript shows how cronolog can be used for such a
test.  Lines prefixed by "**" are the lines I typed.  The log file
template creates a new log file every minute in a new directory.  I
specified a start time a minute before the change from GMT to BST
(British Summer Time).

 ** [andrew@icarus src]$ cronolog -x- -s "29 March 1998 00:59" \
        "tmp/%Y.%m.%d/%H:%M/%Y.%m.%d-%H:%M-access.log"
    cronolog version 1.5b8

    Copyright (C) 1997 Ford & Mason Ltd.
    This is free software; see the source for copying conditions.
    There is NO warranty; not even for MERCHANTABILITY or FITNESS
    FOR A PARTICULAR PURPOSE.

    Written by Andrew Ford <A.Ford@ford-mason.co.uk>

    The latest version can be found at:

	http://www.ford-mason.co.uk/resources/cronolog/

    Using offset of 1606971 seconds from real time
    Determining periodicity of "tmp/%Y.%m.%d/%H:%M/%Y.%m.%d-%H:%M-access.log"
    %Y -> yearly
    %m -> monthly
    %d -> daily
    %H -> hourly
    %M -> per minute
    periodicity = minute
 ** MESSAGE ONE
    1998/03/29-00:59:04 GMT (891133144): using log file \
        "tmp/1998.03.29/00:59/1998.03.29-00:59-access.log" \
        until 1998/03/29-02:00:00 BST (891133200) (for 56 secs)
    Creating missing components of "tmp/1998.03.29/00:59/1998.03.29-00:59-access.log"
    Testing directory "tmp"
    Directory "tmp" does not exist -- creating
    Testing directory "tmp/1998.03.29"
    Directory "tmp/1998.03.29" does not exist -- creating
    Testing directory "tmp/1998.03.29/00:59"
    Directory "tmp/1998.03.29/00:59" does not exist -- creating
    1998/03/29-00:59:04 GMT (891133144): wrote message; \
        next period starts at 1998/03/29-02:00:00 BST (891133200) in 56 secs
 ** MESSAGE TWO
    1998/03/29-00:59:32 GMT (891133172): wrote message; \
        next period starts at 1998/03/29-02:00:00 BST (891133200) in 28 secs
 ** MESSAGE THREE
    1998/03/29-02:00:00 BST (891133200): using log file \
        "tmp/1998.03.29/02:00/1998.03.29-02:00-access.log" \
        until 1998/03/29-02:01:00 BST (891133260) (for 60 secs)
    Creating missing components of "tmp/1998.03.29/02:00/1998.03.29-02:00-access.log"
    Initial prefix "tmp" known to exist
    Initial prefix "tmp/1998.03.29" known to exist
    Testing directory "tmp/1998.03.29/02:00"
    Directory "tmp/1998.03.29/02:00" does not exist -- creating
    1998/03/29-02:00:00 BST (891133200): wrote message; \
        next period starts at 1998/03/29-02:01:00 BST (891133260) in 60 secs

This output shows cronolog:
 * checking the periodicity of the log file template
 * creating missing directories for new log files
 * skipping stat() tests on directories it "knows" already exist
 * recalculating the time until a new log file needs to be opened


NEWS
1.6.1 (20-Dec-1999)

* Cronolog submitted to Apache Software Foundation


1.6 (16-Dec-1999)

* Fixed bug in cronolog whereby if it was started shortly before the end of 
  a period it would not rotate until the end of the next period.
  (Thanks to  Charles Thayer <thayer@mediabridge.com> for identifying the bug 
  and supplying the fix.)


1.5b9 (04-Jun-1998)

* Fixed typo in cronolog.c and cronotest.c  [[ if (start_time); { ... } ]]
  (gcc -Wall doesn't even flag it!)

* changed definition of multi-line string constants to use proper ANSI string 
  concatenation 


1.5b8 (08-Mar-1998)

* Log files only created after a message has been read (you shouldn't
  get empty files)

* time() called only once per message 

* Extra debugging code (see README)

* Updated documentation


Major changes in release 1.5 (11-Apr-1997):

* Fixed problem with change to daylight saving time (summer time).

* Restructured to conform to GNU standards (now uses GNU automake and autoconf).

Major changes in release 1.4 (20-Dec-1996):

* Optimization of create_subdirs: don't bother stat'ing directories
  that form part of a common prefix with the last filename checked.


Major changes in release 1.3 (16-Dec-1996):

* Fixed portability bugs. 


Major changes in release 1.2 (15-Dec-1996):

* Creates missing directories as needed.

* First version announced on comp.infosys.www.servers.unix (renamed to cronolog).

   
Major changes in release 1.1 (09-Dec-1996):

* Fixed problem with log files being created with the wrong dates.

  (When a new log file is due to be created, the time used for filling
  out the template should be the start of the current period and not
  the time at which a new log file is due.  This was observed with daily
  logs where there were no transfers for two days; the next log file
  generated had the date of the first of the days with no transfers.)

* Also added more comments and fixed handling of week numbers where
  weeks start on a Monday.


Major changes in release 1.0 (06-Dec-1996):

* Initial version sent to Apache developers' mailing list 
  (at that time cronolog was called strftimelog)


src/cronoutils.c
/* ====================================================================
 * Copyright (c) 1995-1999 The Apache Group.  All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 *
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer. 
 *
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in
 *    the documentation and/or other materials provided with the
 *    distribution.
 *
 * 3. All advertising materials mentioning features or use of this
 *    software must display the following acknowledgment:
 *    "This product includes software developed by the Apache Group
 *    for use in the Apache HTTP server project (http://www.apache.org/)."
 *
 * 4. The names "Apache Server" and "Apache Group" must not be used to
 *    endorse or promote products derived from this software without
 *    prior written permission. For written permission, please contact
 *    apache@apache.org.
 *
 * 5. Products derived from this software may not be called "Apache"
 *    nor may "Apache" appear in their names without prior written
 *    permission of the Apache Group.
 *
 * 6. Redistributions of any form whatsoever must retain the following
 *    acknowledgment:
 *    "This product includes software developed by the Apache Group
 *    for use in the Apache HTTP server project (http://www.apache.org/)."
 *
 * THIS SOFTWARE IS PROVIDED BY THE APACHE GROUP ``AS IS'' AND ANY
 * EXPRESSED OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
 * PURPOSE ARE DISCLAIMED.  IN NO EVENT SHALL THE APACHE GROUP OR
 * ITS CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
 * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT
 * NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
 * LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT,
 * STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
 * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED
 * OF THE POSSIBILITY OF SUCH DAMAGE.
 * ====================================================================
 *
 * This software consists of voluntary contributions made by many
 * individuals on behalf of the Apache Group and was originally based
 * on public domain software written at the National Center for
 * Supercomputing Applications, University of Illinois, Urbana-Champaign.
 * For more information on the Apache Group and the Apache HTTP server
 * project, please see <http://www.apache.org/>.
 *
 */
/*
 * cronoutils -- utilities for the cronolog program
 *
 * Copyright (c) 1996-1999 by Ford & Mason Ltd
 *
 * This software was submitted by Ford & Mason Ltd to the Apache
 * Software Foundation in December 1999.  Future revisions and
 * derivatives of this source code must acknowledge Ford & Mason Ltd
 * as the original contributor of this module.  All other licensing
 * and usage conditions are those of the Apache Software Foundation.
 *
 * Originally written by Andrew Ford <A.Ford@ford-mason.co.uk>
 *
 */

#include "cronoutils.h"
/*extern char *tzname[2];*/


#ifdef _WIN32
#include "strptime.h"
#endif
/* debug_file is the file to output debug messages to.  No debug
 * messages are output if it is set to NULL. 
 */
FILE	*debug_file = NULL;


/* America and Europe disagree on whether weeks start on Sunday or
 * Monday - weeks_start_on_mondays is set if a %U specifier is encountered.
 */
int	weeks_start_on_mondays = 0;


/* periods[] is an array of the names of the periods.
 */
char	*periods[] = 
{
    "second",
    "minute",
    "hour",
    "day",
    "week",
    "month",
    "year",
    "aeon"	/* i.e. once only */
};

/* period_seconds[] is an array of the number of seconds in a period.
 */
int	period_seconds[] = 
{
    1,
    60,
    60 * 60,
    60 * 60 * 24,
    60 * 60 * 24 * 7,
    60 * 60 * 24 * 31,
    60 * 60 * 24 * 36
};

/* Open a new log file: determine the start of the current
 * period, generate the log file name from the template,
 * determine the end of the period and open the new log file.
 *
 * Returns the file descriptor of the new log file and also sets the
 * name of the file and the start time of the next period via pointers
 * supplied.
 */
int
new_log_file(const char *template, const char *linkname, mode_t linktype, const char *prevlinkname,
	     PERIODICITY periodicity, int period_multiple, int period_delay,
	     char *pfilename, size_t pfilename_len,
	     time_t time_now, time_t *pnext_period)
{
    time_t 	start_of_period;
    struct tm 	*tm;
    int 	log_fd;

    start_of_period = start_of_this_period(time_now, periodicity, period_multiple);
    tm = localtime(&start_of_period);
    strftime(pfilename, BUFSIZE, template, tm);
    *pnext_period = start_of_next_period(start_of_period, periodicity, period_multiple) + period_delay;
    
    DEBUG(("%s (%d): using log file \"%s\" from %s (%d) until %s (%d) (for %d secs)\n",
	   timestamp(time_now), time_now, pfilename, 
           timestamp(start_of_period), start_of_period,
	   timestamp(*pnext_period), *pnext_period,
	   *pnext_period - time_now));
    
    log_fd = open(pfilename, O_WRONLY|O_CREAT|O_APPEND, FILE_MODE);
    
#ifndef DONT_CREATE_SUBDIRS
    if ((log_fd < 0) && (errno == ENOENT))
    {
	create_subdirs(pfilename);
	log_fd = open(pfilename, O_WRONLY|O_CREAT|O_APPEND, FILE_MODE);
    }
#endif	    

    if (log_fd < 0)
    {
	perror(pfilename);
	exit(2);
    }

    if (linkname)
    {
	create_link(pfilename, linkname, linktype, prevlinkname);
    }
    return log_fd;
}

/* Try to create missing directories on the path of filename.
 *
 * Note that on a busy server there may theoretically be many cronolog
 * processes trying simultaneously to create the same subdirectories
 * so ignore any EEXIST errors on mkdir -- they probably just mean
 * that another process got there first.
 *
 * Unless CHECK_ALL_PREFIX_DIRS is defined, we save the directory of
 * the last file tested -- any common prefix should exist.  This
 * probably only saves a few stat system calls at the start of each
 * log period, but it might as well be done. 
 */
void
create_subdirs(char *filename)
{
#ifndef CHECK_ALL_PREFIX_DIRS
    static char	lastpath[MAX_PATH] = "";
#endif
    struct stat stat_buf;
    char	dirname[MAX_PATH];
    char	*p;
    
    DEBUG(("Creating missing components of \"%s\"\n", filename));
    for (p = filename; (p = strchr(p, '/')); p++)
    {
	if (p == filename)
	{
	    continue;	    /* Don't bother with the root directory */
	}
	
	memcpy(dirname, filename, p - filename);
	dirname[p-filename] = '\0';
	
#ifndef CHECK_ALL_PREFIX_DIRS
	if (strncmp(dirname, lastpath, strlen(dirname)) == 0)
	{
	    DEBUG(("Initial prefix \"%s\" known to exist\n", dirname));
	    continue;
	}
#endif

	DEBUG(("Testing directory \"%s\"\n", dirname));
	if (stat(dirname, &stat_buf) < 0)
	{
	    if (errno != ENOENT)
	    {
		perror(dirname);
		exit(2);
	    }
	    else
	    {
		DEBUG(("Directory \"%s\" does not exist -- creating\n", dirname));
		if ((mkdir(dirname, DIR_MODE) < 0) && (errno != EEXIST))
#ifndef _WIN32
		{
#else
                if ((mkdir(dirname) < 0) && (errno != EEXIST))
#endif
		    perror(dirname);
		    exit(2);
		}
	    }
	}
    }
#ifndef CHECK_ALL_PREFIX_DIRS
    strcpy(lastpath, dirname);
#endif
}

/* Create a hard or symbolic link to a filename according to the type specified.
 *
 * This function could do with more error checking! 
 */
void
create_link(char *pfilename, 
	    const char *linkname, mode_t linktype,
	    const char *prevlinkname)
{
    struct stat		stat_buf;
    
    if (stat(prevlinkname, &stat_buf) == 0)
    {
	unlink(prevlinkname);
    }
    if (stat(linkname, &stat_buf) == 0)
    {
	if (prevlinkname) {
	    rename(linkname, prevlinkname);
	}
	else {
	    unlink(linkname);
	}
    }
#ifndef _WIN32
    if (linktype == S_IFLNK)
    {
	symlink(pfilename, linkname);
    }
    else
    {
	link(pfilename, linkname);
    }
#else
    fprintf(stderr, "Creating link from %s to %s not supported", pfilename, linkname);
#endif    
}

/* Examine the log file name specifier for strftime conversion
 * specifiers and determine the period between log files. 
 * Smallest period allowed is per minute.
 */
PERIODICITY
determine_periodicity(char *spec)
{
    PERIODICITY	periodicity = ONCE_ONLY;
    char 	ch;
    
    DEBUG(("Determining periodicity of \"%s\"\n", spec));
    while ((ch = *spec++) != 0)
    {
	if (ch == '%')
	{
	    ch = *spec++;
	    if (!ch) break;
	    
	    switch (ch)
	    {
	    case 'y':		/* two digit year */
	    case 'Y':		/* four digit year */
		if (periodicity > YEARLY)
		{
		    DEBUG(("%%%c -> yearly\n", ch));
		    periodicity = YEARLY;
		}
		break;

	    case 'b':		/* abbreviated month name */
	    case 'h':		/* abbreviated month name (non-standard) */
	    case 'B':		/* full month name */
	    case 'm':		/* month as two digit number (with
				   leading zero) */
		if (periodicity > MONTHLY)
		{
		    DEBUG(("%%%c -> monthly\n", ch));
		    periodicity = MONTHLY;
		}
  	        break;
		
	    case 'U':		/* week number (weeks start on Sunday) */
	    case 'W':		/* week number (weeks start on Monday) */
	        if (periodicity > WEEKLY)
		{
		    DEBUG(("%%%c -> weeky\n", ch));
		    periodicity = WEEKLY;
		    weeks_start_on_mondays = (ch == 'W');
		}
		break;
		
	    case 'a':		/* abbreviated weekday name */
	    case 'A':		/* full weekday name */
	    case 'd':		/* day of the month (with leading zero) */
	    case 'e':		/* day of the month (with leading space -- non-standard) */
	    case 'j':		/* day of the year (with leading zeroes) */
	    case 'w':		/* day of the week (0-6) */
	    case 'D':		/* full date spec (non-standard) */
	    case 'x':		/* full date spec */
	        if (periodicity > DAILY)
		{
		    DEBUG(("%%%c -> daily\n", ch));
		    periodicity = DAILY;
		}
  	        break;
	    
	    case 'H':		/* hour (24 hour clock) */
	    case 'I':		/* hour (12 hour clock) */
	    case 'p':		/* AM/PM indicator */
	        if (periodicity > HOURLY)
		{
		    DEBUG(("%%%c -> hourly\n", ch));
		    periodicity = HOURLY;
		}
		break;
		
	    case 'M':		/* minute */
	        if (periodicity > PER_MINUTE)
		{
		    DEBUG(("%%%c -> per minute\n", ch));
		    periodicity = PER_MINUTE;
		}
		break;
		
	    case 'S':		/* second */
	    case 's':		/* seconds since the epoch (GNU non-standard) */
	    case 'c':		/* full time and date spec */
	    case 'T':		/* full time spec */
	    case 'r':		/* full time spec (non-standard) */
	    case 'R':		/* full time spec (non-standard) */
		DEBUG(("%%%c -> per second", ch));
		periodicity = PER_SECOND;

	    default:		/* ignore anything else */
		DEBUG(("ignoring %%%c\n", ch));
	        break;
	    }
	}
    }
    return periodicity;
}

/* 
 */
PERIODICITY 
parse_timespec(char *optarg, int *p_period_multiple)
{
    PERIODICITY		periodicity	= INVALID_PERIOD;
    int			period_multiple = 1;
    char 		*p = optarg;
    
    /* Skip leading whitespace */
    
    while (isspace(*p)) { p++; }
    
    
    /* Parse a digit string */
    
    if (isdigit(*p)) {
	period_multiple = *p++ - '0';
	
	while (isdigit(*p)) {
	    period_multiple *= 10;
	    period_multiple += (*p++ - '0');
	}
    }

    
    /* Skip whitespace */

    while (isspace(*p)) { p++; }
    
    if (strncasecmp(p, "sec", 3) == 0) {
	if (period_multiple < 60) {
	    periodicity = PER_SECOND;
	}
    }
    else if (strncasecmp(p, "min", 3) == 0) {
	if (period_multiple < 60) {
	    periodicity = PER_MINUTE;
	}
    }
    else if (strncasecmp(p, "hour", 4) == 0) {
	if (period_multiple < 24) {
	    periodicity = HOURLY;
	}
    }
    else if (strncasecmp(p, "day", 3) == 0) {
	if (period_multiple <= 31) {
	    periodicity = DAILY;
	}
    }
    else if (strncasecmp(p, "week", 4) == 0) {
	if (period_multiple < 53) {
	    periodicity = WEEKLY;
	}
    }
    else if (strncasecmp(p, "mon", 3) == 0) {
	if (period_multiple <= 12) {
	    periodicity = MONTHLY;
	}
    }
    *p_period_multiple = period_multiple;
    return periodicity;
}

/* To determine the time of the start of the next period add just
 * enough to move beyond the start of the next period and then
 * determine the time of the start of that period.
 *
 * There is a potential problem if the start or end of daylight saving
 * time occurs during the current period. 
 */
time_t
start_of_next_period(time_t time_now, PERIODICITY periodicity, int period_multiple)
{
    time_t	start_time;
    
    switch (periodicity)
    {
    case YEARLY:
	start_time = (time_now + 366 * SECS_PER_DAY + DST_ALLOWANCE);
	break;

    case MONTHLY:
	start_time = (time_now + 31 * SECS_PER_DAY + DST_ALLOWANCE);
	break;

    case WEEKLY:
	start_time = (time_now + SECS_PER_WEEK + DST_ALLOWANCE);
	break;
	
    case DAILY:
	start_time = (time_now + SECS_PER_DAY + DST_ALLOWANCE);
	break;
	
    case HOURLY:
	start_time = time_now + period_multiple * SECS_PER_HOUR + LEAP_SECOND_ALLOWANCE;
	break;

    case PER_MINUTE:
	start_time = time_now + period_multiple * SECS_PER_MIN + LEAP_SECOND_ALLOWANCE;
	break;

    case PER_SECOND:
	start_time = time_now + 1;
	break;

    default:
	start_time = FAR_DISTANT_FUTURE;
	break;
    }
    return start_of_this_period(start_time, periodicity, period_multiple);
}

/* Determine the time of the start of the period containing a given time.
 * Break down the time with localtime and subtract the number of
 * seconds since the start of the period.  If the length of period is
 * equal or longer than a day then we have to check tht the
 * calculation is not thrown out by the start or end of daylight
 * saving time.
 */
time_t
start_of_this_period(time_t start_time, PERIODICITY periodicity, int period_multiple)
{
    struct tm	tm_initial;
    struct tm	tm_adjusted;
    int		expected_mday;
    
#ifndef _WIN32
    localtime_r(&start_time, &tm_initial);
#else
    struct tm * tempTime;
    
    tempTime = localtime(&start_time);
    if (NULL != tempTime)
    {
        memcpy(&tm_initial, tempTime, sizeof(struct tm));

        free(tempTime);
        tempTime = NULL;
    }
#endif    
    switch (periodicity)
    {
    case YEARLY:
    case MONTHLY:
    case WEEKLY:
    case DAILY:
	switch (periodicity)
	{
	case YEARLY:
	    start_time -= (  (tm_initial.tm_yday * SECS_PER_DAY)
			   + (tm_initial.tm_hour * SECS_PER_HOUR)
			   + (tm_initial.tm_min  * SECS_PER_MIN)
			   + (tm_initial.tm_sec));
	    expected_mday = 1;
	    break;

	case MONTHLY:
	    start_time -= (  ((tm_initial.tm_mday - 1) * SECS_PER_DAY)
			   + ( tm_initial.tm_hour      * SECS_PER_HOUR)
			   + ( tm_initial.tm_min       * SECS_PER_MIN)
			   + ( tm_initial.tm_sec));
	    expected_mday = 1;
	    break;
	    
	case WEEKLY:
	    if (weeks_start_on_mondays)
	    {
		tm_initial.tm_wday = (6 + tm_initial.tm_wday) % 7;
	    }
	    start_time -= (  (tm_initial.tm_wday * SECS_PER_DAY)
			   + (tm_initial.tm_hour * SECS_PER_HOUR)
			   + (tm_initial.tm_min  * SECS_PER_MIN)
			   + (tm_initial.tm_sec));
	    expected_mday = tm_initial.tm_mday;
	    break;
	
	case DAILY:
	    start_time -= (  (tm_initial.tm_hour * SECS_PER_HOUR)
			   + (tm_initial.tm_min  * SECS_PER_MIN )
			   +  tm_initial.tm_sec);
	    expected_mday = tm_initial.tm_mday;
	    break;

	default:
	    fprintf(stderr, "software fault in start_of_this_period()\n");
	    exit(1);
	}

	/* If the time of day is not equal to midnight then we need to
	 * adjust for daylight saving time.  Adjust the time backwards
	 * by the value of the hour, minute and second fields.  If the
	 * day of the month is not as expected one then we must have
	 * adjusted back to the previous day so add 24 hours worth of
	 * seconds.  
	 */
#ifndef _WIN32
	localtime_r(&start_time, &tm_adjusted);    
#else
	tempTime = localtime(&start_time);
	if (NULL != tempTime)
	{
	    memcpy(&tm_adjusted, tempTime, sizeof(struct tm));
	    
	    free(tempTime);
	    tempTime = NULL;
	}
#endif    
	if (   (tm_adjusted.tm_hour != 0)
	    || (tm_adjusted.tm_min  != 0)
	    || (tm_adjusted.tm_sec  != 0))
	{
	    char	sign   = '-';
	    time_t      adjust = - (  (tm_adjusted.tm_hour * SECS_PER_HOUR)
				    + (tm_adjusted.tm_min  * SECS_PER_MIN)
				    + (tm_adjusted.tm_sec));

	    if (tm_adjusted.tm_mday != expected_mday)
	    {
		adjust += SECS_PER_DAY;
		sign = '+';
	    }
	    start_time += adjust;

	    if (adjust < 0)
	    {
		adjust = -adjust;
	    }
	    
	    DEBUG(("Adjust for dst: %02d/%02d/%04d %02d:%02d:%02d -- %c%0d:%02d:%02d\n",
		   tm_initial.tm_mday, tm_initial.tm_mon+1, tm_initial.tm_year+1900,
		   tm_initial.tm_hour, tm_initial.tm_min,   tm_initial.tm_sec, sign,
		   adjust / SECS_PER_HOUR, (adjust / 60) % 60, adjust % SECS_PER_HOUR));
	}
	break;

    case HOURLY:
	start_time -= (tm_initial.tm_sec + tm_initial.tm_min * SECS_PER_MIN);
	if (period_multiple > 1) {
	    start_time -= SECS_PER_HOUR * (tm_initial.tm_hour -
					   period_multiple * (tm_initial.tm_hour / period_multiple));
	}
	break;

    case PER_MINUTE:
	start_time -= tm_initial.tm_sec;
	if (period_multiple > 1) {
	    start_time -= SECS_PER_MIN * (tm_initial.tm_min -
					  period_multiple * (tm_initial.tm_min / period_multiple));
	}
	break;

    case PER_SECOND:	/* No adjustment needed */
    default:
	break;
    }
    return start_time;
}

/* Converts struct tm to time_t, assuming the data in tm is UTC rather
 * than local timezone (as mktime assumes).
 *  
 * Contributed by Roger Beeman <beeman@cisco.com>. 
 */
time_t
mktime_from_utc(struct tm *t)
{
   time_t tl, tb;

   tl = mktime(t);
   tb = mktime(gmtime(&tl));                               
   return (tl <= tb ? (tl + (tl - tb)) : (tl - (tb - tl)));
}

/* Check whether the string is processed well.  It is processed if the
 * pointer is non-NULL, and it is either at the `GMT', or at the end
 * of the string.
 */
static int
check_end(const char *p)
{
   if (!p)
      return 0;
   while (isspace(*p))
      ++p;
   if (!*p || (p[0] == 'G' && p[1] == 'M' && p[2] == 'T'))
      return 1;
   else
      return 0;
}

/* NOTE: We don't use `%n' for white space, as OSF's strptime uses
   it to eat all white space up to (and including) a newline, and
   the function fails (!) if there is no newline.

   Let's hope all strptime-s use ` ' to skipp *all* whitespace
   instead of just one (it works that way on all the systems I've
   tested it on). */

static char *european_date_formats[] = 
{	
    "%d %b %Y %T",       	/* dd mmm yyyy HH:MM:SS */
    "%d %b %Y %H:%M",       	/* dd mmm yyyy HH:MM    */
    "%d %b %Y",       		/* dd mmm yyyy          */
    "%d-%b-%Y %T",       	/* dd-mmm-yyyy HH:MM:SS */
    "%d-%b-%Y %H:%M",       	/* dd-mmm-yyyy HH:MM    */
    "%d-%b-%y %T",       	/* dd-mmm-yy   HH:MM:SS */
    "%d-%b-%y %H:%M",  		/* dd-mmm-yy   HH:MM    */
    "%d-%b-%Y",
    "%b %d %T %Y",
    "%b %d %Y",
    NULL
};

static char *american_date_formats[] = 
{	
    "%b %d %Y %T",       	/* dd mmm yyyy HH:MM:SS */
    "%b %d %Y %H:%M",       	/* dd mmm yyyy HH:MM    */
    "%b %d %Y",       		/* dd mmm yyyy          */
    "%b-%d-%Y %T",       	/* dd-mmm-yyyy HH:MM:SS */
    "%b-%d-%Y %H:%M",       	/* dd-mmm-yyyy HH:MM    */
    "%b-%d-%Y",
    "%b/%d/%Y %T",
    "%b/%d/%Y %H:%M",
    "%b/%d/%Y",
    NULL
};



time_t
parse_time(char *time_str, int use_american_date_formats)
{
   struct tm    tm;
   char		**date_formats;
   
   memset(&tm, 0, sizeof (tm));
   tm.tm_isdst = -1;
   
   
   for (date_formats = (use_american_date_formats
			? american_date_formats
			: european_date_formats);
	*date_formats;
	date_formats++)
   {
       if (check_end((const char *)strptime(time_str, *date_formats, &tm)))
	   return mktime_from_utc(&tm);
   }
   
   return -1;
}
  


/* Simple debugging print function.
 */
void
print_debug_msg(char *msg, ...)
{
    va_list	ap;
    
    va_start(ap, msg);
    vfprintf(debug_file, msg, ap);
}


/* Build a timestamp and return a pointer to it.
 * (has a number of static buffers that are rotated).
 */ 
char *
timestamp(time_t thetime)
{
    static int	index = 0;
    static char	buffer[4][80];
    struct tm	*tm;
    char	*retval;

    retval = buffer[index++];
    index %= 4;
    
    tm = localtime(&thetime);
    strftime(retval, 80, "%Y/%m/%d-%H:%M:%S %Z", tm);
    return retval;
}

    
